Rsrlib/tmp/Docs/tu.1
.\" bf thingy 
.\" Copyright (c) 1994,1995,1996, Handmade Software Inc.
.\"
.TH tu 1 "08 April 1996
.SH NAME   
\fBtu\fP \- unpack binary RSR download files to ASCII
.SH SYNOPSIS
\fBtu\fP [\fBoptions\fP] filename [\fBoptional additional filenames\fP]
.SH AVAILABILITY
Versions of \fBtu\fP are available for most \fBUnix\fP systems,
as well as PCs running \fBDOS\fP.
.SH DESCRIPTION
\fItu\fP converts files downloaded from a Yankee/PNL rotating shadowband
radiometer into ASCII formatted data.
\fItu\fP can also optionally calibrate and cosine correct its input.
Download files of every type
including \fIType 2\fP are accepted.
When more than one download filenames
are given, the output is concatenated into a single output file.
.PP
The appearance of the output file is determined by several
parameters which are preset to default values, but which may be
overridden by both configuration file values and command line
flags. Values in the configuration file override default values,
and command line flags take precedence over all others.
.SH CONFIGURATION FILE
The default configuration filename is ".turc" on Unix systems, and
"tu.ini" on DOS. The configuration file can also be selected with
the "-f" flag.
.PP
The configuration file is composed of lines of the form
.br
name = value
.br
The effect is to set the parameter \fIname\fP to \fIvalue\fP.
Comments are introduced with the '#' character - everything from
the '#' to the end of the current line is ignored, as are blank lines.
The recognized parameter names are given below, along with their
corresponding command line equivalents.
.SH COMMAND LINE FLAGS
Flags are indicated on the command line by a '-' character followed
immediately by a single character (the \fBflag\fP) which indicates
a parameter to set. Most flags are also followed by a value to which
the flag quantity is set. However, two flags have no configuration file
equivalent, and no following value.
.TP
-h
Help. Print a brief usage summary then exit.
.TP
-V
Version. Print the version of this program then exit.
.PP
One flag has no config file equivalent, but requires an argument.
.TP
-f \fIfilename\fP
\fBfilename\fP indicates an alternate configuration file. If the
file does not exist or can't be opened, the program exits with failure.
.PP
The remaining flags all have a required argument, and an equivalent
configuration file name. The config file name is shown in parenthesis.
.TP
-v \fIverbosity\fP (verbosity)
Set the program chattiness to 'quiet', 'normal' (the default) or 'verbose'.
When set to 'verbose,' the program prints some RSR header information,
as well as the values of the runtime configuration parameters.
When 'normal,' only serious errors and selected informational
messages are displayed. When 'quiet,' no messages at all are given.
This implies that the success or failure of the conversion must
be determined from the calling environment.
.TP
-z \fItimezone\fP (timezone)
The number of hours (west positive) from Greenwich for this file.
Any floating point number is accepted. The default is 0.0 (GMT).
.TP
-n \fInightstring\fP (nighttime)
RSR 'daytime only' channels are represented by this value at night.
The default is '-9999'. The special string 'omit' is recognized, and
causes daytime-only channels to be omitted from the output at night.
.TP
-s \fIseparator\fP (separator)
Fields in the output file are separated by this string. The special
string 'format' is recognized and causes tabular (fixed column width)
output.
.TP
-d \fIdatefmt\fP (date)
The date prepended to each record is printed according to \fIdatefmt\fP:
.br
\fIdatefmt\fP\h'|1i'\fIStyle\fP
.br
doy0\h'|1i'yy doy (Jan 1 == 0)
.br
doy1\h'|1i'yy doy (Jan 1 == 1; this is the default)
.br
mdy\h'|1i'mm dd yy
.br
dmy\h'|1i'dd mm yy
.br
unix\h'|1i'seconds since 1/1/1970
.br
rsr\h'|1i'days since 1/1/1900
.br
joe\h'|1i'(days since 1900).(fractional day)
.TP
-t \fItimefmt\fP (time)
The time printed for each record is formatted according to \fItimefmt\fP:
.br
\fItimefmt\fP\h'|1i'\fIStyle\fP
.br
hms\h'|1i'hh mm ss
.br
hours\h'|1i'hours.fractional-hour (this is the default)
.br
daysec\h'|1i'seconds since midnight
.br
\fBNote: if the date format is 'unix' or 'joe'
then no time field is produced.\fP
.TP
-o \fIoutputfile\fP (outfile)
The name of the file to write the ASCII data to. The default is the
screen (stdout).
.TP
-g \fIgapval\fP (gap)
Data fields are represented by \fIgapval\fP during 'gap records.'
.TP
-H \fIdohead\fP (header)
Allowed values are 'no' (the default) and 'yes'. If 'yes', the first
line in the output file will show the RSR header parameters.
.TP
-r \fIinterval\fP (reptime)
Indicate the part of the sample interval that is reported with each
record. 'end' (the default) reports the record time as of the end
of the sample interval, 'middle' reports the midpoint, and 'start'
reports the beginning.
.TP
-c \fIcoscorfile\fP (coscor)
If this flag is present in the config file or command line, the indicated
\fIcoscorfile\fP is assumed to be a \fISolarInfo\fP-formatted file
that should be opened, read, applied
to relevant fields before being output. If not
given, no cosine correction is applied.
.TP
-l \fIcalibfile\fP (calibration)
If this flag is present in the config file or command line, the named
\fIcalibfile\fP is assumed to contain a list of calibration functions
to be applied to the RSR data. The format of the calibration file is
described in the file "Tu-calib.txt". If the named file is missing or
malformed, \fItu\fP will indicate the source of the problem
and attempt to carry on.
.TP
-x \fIextras\fP (extra)
\fItu\fP can produce output fields not present in the input file. These
are solar geometry-related quantities that may be useful in conjunction
with the raw data. \fIextra\fP consists of a string of letters, each of
which indicates that a solar quantity should be produced in the output
file. The fields are produced \fIafter\fP the RSR data, and in the order
given in the string.
If relevant, a lower-case character indicates that the field
should be printed in degrees, and upper-case indicates radians.
Otherwise, upper- and lower-case may be used interchangeably.
.br
\fICharacter\fP\h'|1i'\fIQuantity\fP
.br
a/A\h'|1i'Solar azimuth
.br
e/E\h'|1i'Solar elevation
.br
z/Z\h'|1i'Solar zenith angle
.br
h/H\h'|1i'Hour angle
.br
d/D\h'|1i'Declination
.br
m/M\h'|1i'Air mass
.br
s/S\h'|1i'Relative solar distance
.br
t/T\h'|1i'True solar time
.br
\fBNote: Air mass is not defined when elevation < 0; therefore, airmass
obeys the same rules as 'daytime only' data. All others are treated
as 'all the time' quantities.
.SH EXAMPLES
The following command reads the file named \fIrsr.dat\fP and writes
converted ASCII data to a file called \fIrsr.txt\fP. The timezone is
US EST, and the cosine correction file is \fIsi.dat\fP.
.PP
\fCtu -o rsr.txt -z 5 -c si.dat rsr.dat\fP
.PP
A configuration which changes some of the default
values is shown next:
.PP
.nf
\fC
verbosity\h'|1i'= verbose
timezone\h'|1i'= 8	# PST
nighttime\h'|1i'= NA
date\h'|1i'= doy1
time\h'|1i'= hms
outfile\h'|1i'= rsr-out.dat
gap\h'|1i'= -1
reptime\h'|1i'= middle
\fP
.fi
.SH SEE ALSO
convert(1), rsrsplit(1), jd(1), ph(1)
.SH FILES
\&.turc (Unix)
.br
tu.ini (DOS)
.SH AUTHOR
Handmade Software, Inc. Send bug reports, comments, and flames
to \fIrob@dinner.asrc.albany.edu\fP and/or \fIjim@grunt.asrc.albany.edu\fP.
.SH BUGS
.TP
Fixed as reported. Send information to the address shown above.


tu.c
#include <stdio.h>
#include <stdlib.h>

#include "tu.h"

/*
** 'The Unpacker'
**
** This program reads one or more RSR dump files and emits an
** ASCII version of the same.  This version is based
** on a portable unpacking library developed by
** Handmade Software.
**
** November 11, 1994, Handmade Software Inc.
** version 1.00
**
** 12/14/94 rs
** Add -c/coscor processing parameter that indicates a file to use
** for cosine-correcting the data. If none given, no CC is done. If
** a filename is provided which is in any way incorrect, processing stops.
** version 1.1
**
** 5/3/95 rs
** Remove the '-q' flag and replace its functionality with a new
** option to the 'verbosity' flag/config: 'quiet'. The command line
** now accepts '-v quiet' which will act just like '-q' used to.
** 'quiet' is also accepted in the config file.
**
** 5/12/95 rs
** When the 'header output' is selected, tu would not distinguish
** between single-channel and multifilter files. Lee changed the
** meaning of 2 bytes in the header slightly, and so I changed
** emit_header() in unpack.c to output the header slightly differently
** depending on the input file.
**
** 6/26/95 rs
** 'setup_pp' has always bugged me; working on a replacement; if accepted,
** version will be 1.2.2
**
** 8/1/95 js 
** Code in rsr_rec_length() in rsrlibc.c was incorrectly calculating
** the recordlength anytime the number of channels was a multiple
** of 8.  i.e., when nc=8, record length should be 13, old code
** calculated 14. Version 1.2.3
**
** 8/2/95 rs
** Negative values in type 2 files computed incorrectly in library.
**
** 10/27/95 rs
** The revised 'setup_pp' incorrectly processed "extra flags". Fixed.
** A couple of things (uninit corr_kind and 0 or 1 key) in Solarinfo fixed.
** Version 1.2.5
**
** 1/24/96 rs
** Embarking on a major change to incorporate calibration.
** The plan is to add a command line flag and corresponding
** config file entry which indicates that the calibration is to be done,
** and the file to find the calib. data in.
** See the file "calib.txt" for a more
** complete technical description of the method. When complete, a
** major version change is expected, probably 1.3...
**
** 3/20/96 (solstice!) rs
** Getting reasonable numbers from the calibration code. Will need
** tuning and bullet-protection, but so far it's promising. Making
** a version change to 1.2.901.
**
** 4/29/96 rs
** Got a good report from a user of calibration. Also a bug report
** concerning the "date=rsr" format. See 'unpack.c' for the details.
** version 1.2.902.
**
** 6/24/96 rsr
** Re-fixed (?) a bug in mystrtod.c. 1.2.903
**
** 9/19/96 rs
** A bit of a miscalculation in 'do_unpack' about when "fmt4" and "sstoa"
** should be called. Fixed. 1.2.904
**
** 10/24/96 rs
** Changed the order that calibration and cosine correction are called.
** Fixed problem in calibration which referred to raw counts instead of
** corrected (direct, cosine, etc) data. 1.2.905
**
** 10/24/96 rs
** Compiling under DJGPP gives core dump when run as "tu -h"; some
** member variables in pp are unitialized, but this is fixed by setting
** to NULL in ini_pp.
** Also pp->fnms was uninitialized which can cause a core dump when
** freeing if there were no files to unpack. v 1.2.906
**
** 10/29/96 rs
** There was a test in cosine_correct which made sure that direct was
** high enough to do something with; unfortuantely, after calibrating,
** that number may be very small. changed the limit from 1 to 0.0001.
** v 1.2.907
**
** 11/7/96 rs
** Change to interpretation of "error byte" in librsr.
** v 1.2.908
**
** 6/29/98 js
** No change to code, just a reordering of Makefiles and version number stuff.
** Now, version numbers for tar files are keyed of the #define VERSION.  I'm
** trying to make this standard for all ASRC Solar group software (that comes
** from me, anyway).
** v 1.2.909
**
** 7/20/98 js
** in cos.c commented out code that swithced global and beam data.  Rob must have
** done this to make things orthogonal and then forgot he had done it becuase as
** it turns out, the cosine correction code assumes the b d g ordering for single
** channel units.  Thus, the swapping was screwing things up royally.
** v 1.2.910
**
** 2/22/99 js
** fixed dumb but deadly bug in rsrlibc.c.  It involved confusing rf->rec_len_day with 
** rf->rec_n_day in rsr_next_record().  I'm not sure how tu could have worked properly
** at all with this bug in place.
**
** 2/20/2000 js
** fixed 2000 leap year problem in sunae.c
** v 1.2.912
**
** 8/9/2000 js
** changed the emit_header() routine to spit out hex values where it makes more sense
** i.e., flags, aux, all-the-time, etc.
** v 1.2.913
*/

#define VERSION "1.2.913"

char *Version = VERSION;

/*
** the overall processing parameters are described in the
** global variable 'pp'
proc_params	pp;
*/

Verbosity	do_msg_output = Unset;
extern long	timezone;
#ifndef sparc
extern int	daylight;
#endif

int main(int argc, char **argv)
{
    proc_params	*pp,
		*setup_pp(int, char **);
    int		rv,
		do_unpack(proc_params *);
    void	usage(char *, proc_params *);

    timezone = 0L;
#ifndef sparc
    daylight = 0;
#endif

    /* read default, cfg file, and command line parameters */
    if ((pp = setup_pp(argc, argv)) == NULL)
	exit(1);

    /* help? */
    if (pp->help.u_val.ival) {
	usage(argv[0], pp);
	pp_delete(pp);
	return 0;
    }

    /* version? */
    if (pp->version.u_val.ival) {
	char	buf[128];

	sprintf(buf, "%s version %s", argv[0], pp->version.value);
	tu_msg(buf);

	pp_delete(pp);
	return 0;
    }

    rv = do_unpack(pp);

    pp_delete(pp);

    return rv;
}



void tu_advise(char *msg)
{
    if (do_msg_output == Verbose)
	tu_msg(msg);
}


void tu_msg(char *msg)
{
    if (do_msg_output != Quiet && do_msg_output != QuietCmdLine) {
	fputs(msg, stderr);
	fputc('\n', stderr);
    }
}




void pp_delete(proc_params *pp)
{
    param	*p[16];
    int		i;

    if (pp == NULL)
	return;

    p[0] = &pp->timezone;
    p[1] = &pp->cfg_fnm;
    p[2] = &pp->verbosity;
    p[3] = &pp->help;
    p[4] = &pp->version;
    p[5] = &pp->nighttime;
    p[6] = &pp->separator;
    p[7] = &pp->date_fmt;
    p[8] = &pp->time_fmt;
    p[9] = &pp->report_time;
    p[10] = &pp->out_fnm;
    p[11] = &pp->gap;
    p[12] = &pp->extra;
    p[13] = &pp->coscor_fnm;
    p[14] = &pp->header;
    p[15] = &pp->calibration;

    for (i=0; i<16; i++)
	if (p[i] != NULL) {
	    if (p[i]->parameter != NULL) free(p[i]->parameter);
	    if (p[i]->description != NULL) free(p[i]->description);
	    if (p[i]->def_value != NULL) free(p[i]->def_value);
	    if (p[i]->cfg_name != NULL) free(p[i]->cfg_name);
	    if (p[i]->value != NULL) free(p[i]->value);
	    /*
	    ** do this for strings and extra only!
	    if (p[i]->u_val != NULL) free(p[i]->u_val);
	    */
	}

    if (pp->fnms != NULL) {
	for (i=0; i<pp->n_files_to_unpack; i++)
	    if (pp->fnms[i] != NULL)
		free(pp->fnms[i]);
	free(pp->fnms);
    }

    free(pp);
}




/*
** usage()
** print usage message
*/
void usage(char *exe, proc_params *pp)
{
    char	buf[512];
    param	*params[17], **p;

    sprintf(buf, "Usage: %s [flags] fnm [fnm...]", exe);
    tu_msg(buf);
    tu_msg("Unpack the RSR files 'fnm'...\nFlags:");

    params[0] = &pp->timezone;
    params[1] = &pp->cfg_fnm;
    params[2] = &pp->verbosity;
    params[3] = &pp->help;
    params[4] = &pp->version;
    params[5] = &pp->nighttime;
    params[6] = &pp->separator;
    params[7] = &pp->date_fmt;
    params[8] = &pp->time_fmt;
    params[9] = &pp->report_time;
    params[10] = &pp->out_fnm;
    params[11] = &pp->gap;
    params[12] = &pp->extra;
    params[13] = &pp->coscor_fnm;
    params[14] = &pp->header;
    params[15] = &pp->calibration;
    params[16] = NULL;

    for (p=params; *p != NULL; p++) {
	sprintf(buf, "-%c: %s", (*p)->cmdln_char, (*p)->description);
	tu_msg(buf);
    }
}





